crates/weaver_resolved_schema/src/lib.rs

```diff
@@ -64,44 +70,367 @@ pub struct ResolvedTelemetrySchema {
     pub fn stats(&self) -> Stats {
-        let mut registry_stats = Vec::new();
-        for registry in self.registries.values() {
-            registry_stats.push(registry.stats());
-        }
+        let registry_stats = vec![self.registry.stats()];
         Stats {
-            registry_count: self.registries.len(),
             registry_stats,
             catalog_stats: self.catalog.stats(),
         }
     }
+
+    /// Get the attributes of the resolved telemetry schema.
+    #[must_use]
+    pub fn attribute_map(&self) -> HashMap<&str, &Attribute> {
+        self.registry
+            .groups
+            .iter()
+            .filter(|group| group.r#type == GroupType::AttributeGroup)
+            .flat_map(|group| {
+                group.attributes.iter().map(|attr_ref| {
+                    // An attribute ref is a reference to an attribute in the catalog.
+                    // Not finding the attribute in the catalog is a bug somewhere in
+                    // the resolution process. So it's fine to panic here.
+                    let attr = self
+                        .catalog
+                        .attribute(attr_ref)
+                        .expect("Attribute ref not found in catalog. This is a bug.");
+                    (attr.name.as_str(), attr)
+                })
+            })
+            .collect()
+    }
+
+    /// Get the metrics of the resolved telemetry schema.
+    #[must_use]
+    pub fn groups(&self, group_type: GroupType) -> HashMap<String, &Group> {
+        self.registry
+            .groups
+            .iter()
+            .filter(|group| group.r#type == group_type)
+            .map(|group| match group_type {
+                GroupType::AttributeGroup
+                | GroupType::Event
+                | GroupType::Span
+                | GroupType::Resource
+                | GroupType::MetricGroup
+                | GroupType::Scope => (group.name.clone().unwrap_or_default(), group),
+                GroupType::Metric => (group.metric_name.clone().unwrap_or_default(), group),
+            })
+            .collect()
+    }
+
+    /// Generate a diff between the current schema (must be the most recent one)
+    /// and a baseline schema.
+    #[must_use]
+    pub fn diff(&self, baseline_schema: &ResolvedTelemetrySchema) -> SchemaChanges {
+        let mut changes = SchemaChanges::new();
+
+        if let Some(ref manifest) = self.registry_manifest {
+            changes.set_head_manifest(weaver_version::schema_changes::RegistryManifest {
+                semconv_version: manifest.semconv_version.clone(),
+            });
+        }
+
+        if let Some(ref manifest) = baseline_schema.registry_manifest {
+            changes.set_baseline_manifest(weaver_version::schema_changes::RegistryManifest {
+                semconv_version: manifest.semconv_version.clone(),
+            });
+        }
+
+        // Attributes in the registry
+        self.diff_attributes(baseline_schema, &mut changes);
```

Comment:
I have two concerns here:

1. overall attribute renames rely on the *registry concept*.  We can only look at renames/added/removed from attribute groups there.  We should probably avoid looking at deprecated on a per-signal basis *in that instance*.
2. On a per-signal basis, we should be tracking added/removed attributes. (We can ignore changed, as those should be caught in overall "registry" renames.


Example:

- There are two attributes `error.type` and `http.status_code`.
- There is an event `http.my.count`.  It currently uses `error.type`.
- Example1: `http.my.count` *event* is updated to use `http.status_code` instead, using some kind of "ref"/"deprecated" combo.
- Example2: `http.my.count` updates `error.type` from `opt-in` to `recommended`.

Perhaps we don't want to handle this yet.  Curious on @lmolkova's thoughts.

Reply:
Concerns 1 and 2 make sense to me. I can remove rename tracking for everything except the attributes in the registry.

Regarding the examples, could you describe more precisely what you’d like to see in the diff?

Reply:
I think this PR detects renames only (for attributes, metrics, etc) and does not analyze what's changed within a group. 

I can see cases when we want to go deeper, but they are beyond changelog/schema transformation automation. E.g. we do it in compatibility policy checks.

I wonder if good old weaver.yaml + JQ + jinja with access to baseline and new semconv at the same time can cover all possible complex diff scenarios one would want to do.

Reply:
I think at some point we'll want diffs on what to include in a metric or span.  As Lidumila calls out we can defer that until later .

Reply:
Point 1 is fixed in https://github.com/open-telemetry/weaver/pull/400/commits/dbe4316822895918eeadbe2459f4bc46fa30c653.

Point 2 is postponed in a future PR.